src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -863,13 +870,23 @@ public static object DeepCopyInner(object original)
             {
                 copy = copier(original);
                 SerializationContext.Current.RecordObject(original, copy);
+                return copy;
             }
-            else
+
+            var externalSerializer = externalSerializers.FirstOrDefault(serializer => serializer.IsSupportedType(t));
```

Comment:
@sergeybykov A couple use cases for multiple serializers
- Small data blocks use fast non-compressing serializers and big data blocks use compression
- For security requirements, some data blocks may need to be encrypted for transfer over the wire
I'm not sure if those are legitimate or supported use cases.

You're correct about the HashSet<T> and the ambiguity issue with multiple serializers.